### PR TITLE
out-of-body-experience: avatar can walk up steps with less physics lag

### PR DESCRIPTION
**Pull request by AndrewMeadows**
* magic sky hook for avatar when walking up steps
* no expensive sweep tests

### CODE REVIEW

**Comment by AndrewMeadows** [comment]
This PR doesn't yet solve the "driving HMD can pass through walls" problem, but it does eliminate the source of lag in the out-of-body-experience branch when trying to walk near very complex objects.

File: interface/src/avatar/MyCharacterController.cpp [hunk 1]
```diff
@@ -37,12 +37,32 @@ void MyCharacterController::updateShapeIfNecessary() {
...
+                btVector3 yAxis = btVector3(0.0f, 1.0f, 0.0f);
+                btVector3 zAxis = btVector3(0.0f, 0.0f, 1.0f);
+                points[0] = _halfHeight * yAxis;
+                points[1] = -_halfHeight * yAxis;
+                points[2] = (0.75f * _halfHeight) * yAxis - (0.1f * _radius) * zAxis;
+                points[3] = (0.75f * _halfHeight) * yAxis + (0.1f * _radius) * zAxis;
+                points[4] = (0.75f * _halfHeight) * yAxis - (0.1f * _radius) * xAxis;
+                points[5] = (0.75f * _halfHeight) * yAxis + (0.1f * _radius) * xAxis;
+                btCollisionShape* shape = new btConvexHullShape(reinterpret_cast<btScalar*>(points), NUM_POINTS);
+                shape->setMargin(_radius);
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyCharacterController.cpp]: This is my clever fix for the "avatar can walk up walls" problem.  The avatar now uses a ConvexHull that doesn't suffer from the unfortunate "single contact manifold down near the feet" problem against vertical walls.

File: libraries/physics/src/CharacterController.cpp [hunk 2]
```diff
@@ -336,10 +317,25 @@ void CharacterController::playerStep(btCollisionWorld* collisionWorld, btScalar
...
+            _rigidBody->setWorldTransform(transform);
+
+            // make sure the upward velocity is large enough to clear the very top of the step
+            const btScalar MAGIC_STEP_OVERSHOOT_SPEED_COEFFICIENT = 0.5f;
+            minStepUpSpeed = MAGIC_STEP_OVERSHOOT_SPEED_COEFFICIENT * sqrtf(fabsf(2.0f * _gravity * _minStepHeight));
+            btScalar vDotUp = velocity.dot(_currentUp);
+            if (vDotUp < minStepUpSpeed) {
+                velocity += (minStepUpSpeed - vDotUp) * _stepUpVelocity;
+            }
+        }
```
Comment (AndrewMeadows) [inline libraries/physics/src/CharacterController.cpp]: Most of this work was restoration from a previous state of the branch.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 72ad974b3e5c6520a229911b5606768b9068a8ed available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-72ad974.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-72ad974.dmg

**Comment by ZappoMan** [comment]
```
 PhysicsEngine.cpp
    33>C:\jenkins\workspace\beta_pr-master\label\windows\libraries\physics\src\CharacterGhostShape.cpp(15): error C3861: 'assert': identifier not found [C:\jenkins\workspace\beta_pr-master\label\windows\build\libraries\physics\physics.vcxproj]
    33>C:\jenkins\workspace\beta_pr-master\label\windows\libraries\physics\src\CharacterGhostShape.cpp(16): error C3861: 'assert': identifier not found [C:\jenkins\workspace\beta_pr-master\label\windows\build\libraries\physics\physics.vcxproj]
    33>C:\jenkins\workspace\beta_pr-master\label\windows\libraries\physics\src\CharacterGhostShape.cpp(17): error C3861: 'assert': identifier not found [C:\jenkins\workspace\beta_pr-master\label\windows\build\libraries\physics\physics.vcxproj]
         PhysicsLogging.cpp
```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 515b8ef3956c5a370c9f8974b10ecd136a6194ea available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-515b8ef.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-515b8ef.dmg

File: interface/src/avatar/MyCharacterController.cpp [hunk 3]
```diff
@@ -76,3 +90,215 @@ void MyCharacterController::updateShapeIfNecessary() {
...
+                btVector3 localRayStart = localRayEnd - reach;
+                MeOnlyResultCallback result(_rigidBody);
+			    world->rayTest(position + rotation * localRayStart, position + rotation * localRayEnd, result);
+			    if (result.m_closestHitFraction < 1.0f) {
+                    _bottomPoints.push_back(localRayStart + result.m_closestHitFraction * reach);
+                }
+            }
+        }
+    }
+}
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyCharacterController.cpp]: Adding this "ray shotgun" code to help with driving the HMD through the space with collisions against physical geometry.  It isn't actually used for that yet, but it is used for preventing the avatar from walking up really steep slopes.

File: interface/src/avatar/MyCharacterController.cpp [hunk 4]
```diff
@@ -76,3 +90,215 @@ void MyCharacterController::updateShapeIfNecessary() {
...
+    points[0] = _halfHeight * yAxis;
+    points[1] = -_halfHeight * yAxis;
+    points[2] = (0.75f * _halfHeight) * yAxis - (0.1f * _radius) * zAxis;
+    points[3] = (0.75f * _halfHeight) * yAxis + (0.1f * _radius) * zAxis;
+    points[4] = (0.75f * _halfHeight) * yAxis - (0.1f * _radius) * xAxis;
+    points[5] = (0.75f * _halfHeight) * yAxis + (0.1f * _radius) * xAxis;
+    btConvexHullShape* shape = new btConvexHullShape(reinterpret_cast<btScalar*>(points), NUM_POINTS);
+    shape->setMargin(_radius);
+    return shape;
+}
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyCharacterController.cpp]: This is a clever trick for helping the avatar not walk up vertical walls -- it forces the "single manifold point" problem to typically happen at points higher than the _maxStepHeight.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6f5f6d2493223fb0e2b8303486933397efdcf870 available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-6f5f6d2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-6f5f6d2.dmg

File: interface/src/avatar/MyCharacterController.cpp [hunk 5]
```diff
@@ -76,3 +90,215 @@ void MyCharacterController::updateShapeIfNecessary() {
...
+            if (i%2 == 1) {
+                // odd rows have one less point and start a halfStep closer
+                maxJ -= 1;
+                offsetX += 0.5f * stepX;
+            }
+            for (int32_t j = 0; j < maxJ; ++j) {
+                btVector3 localRayEnd(offsetX + (btScalar)(j) * stepX, divisionLine + (btScalar)(i) * stepY, 0.0f);
+                btVector3 localRayStart = localRayEnd - reach;
+                MeOnlyResultCallback result(_rigidBody);
+			    world->rayTest(position + rotation * localRayStart, position + rotation * localRayEnd, result);
```
Comment (hyperlogic) [inline interface/src/avatar/MyCharacterController.cpp]: indent

File: interface/src/avatar/MyCharacterController.cpp [hunk 6]
```diff
@@ -76,3 +90,215 @@ void MyCharacterController::updateShapeIfNecessary() {
...
+            if (i%2 == 1) {
+                // odd rows have one less point and start a halfStep closer
+                maxJ -= 1;
+                offsetX += 0.5f * stepX;
+            }
+            for (int32_t j = 0; j < maxJ; ++j) {
+                btVector3 localRayEnd(offsetX + (btScalar)(j) * stepX, (divisionLine - slop) - (btScalar)(i) * stepY, 0.0f);
+                btVector3 localRayStart = localRayEnd - reach;
+                MeOnlyResultCallback result(_rigidBody);
+			    world->rayTest(position + rotation * localRayStart, position + rotation * localRayEnd, result);
```
Comment (hyperlogic) [inline interface/src/avatar/MyCharacterController.cpp]: indent

**Comment by hifi-gustavo** [comment]
Build console output for commit: ecc6cb49aefca5afee63dca5b3b5ffcd99c6c409 available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-ecc6cb4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-ecc6cb4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5a04c7ceee1725fe84708deae558d7f81c4ff707 available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-5a04c7c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-5a04c7c.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a4308f08a42b177b245a698e4b38394d11275b03 available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-a4308f0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-a4308f0.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6a445a0a9856808dc8ca0a822638c436802e56c0 available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-6a445a0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-6a445a0.dmg

**Comment by hyperlogic** [comment]
Crash on exit:  https://www.bugsplat.com/individualCrash/?id=6870&database=interface_alpha

**Comment by hifi-gustavo** [comment]
Build console output for commit: f414ffaf98815f208c62eb5cb095d224f25dae56 available at: https://gist.github.com/d0b5c8772645d435a4e0c3d8de899520<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-f414ffa.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8822-f414ffa.dmg

**Pull request merged**